pkg/cloudprovider/controller/nodecontroller.go

```diff
@@ -402,49 +361,145 @@ func (s *NodeController) checkNodeReady(node *api.Node) *api.NodeCondition {
+// a long period of time.
+func (s *NodeController) MonitorNodeStatus() error {
+	nodes, err := s.kubeClient.Nodes().List()
 	if err != nil {
 		return err
 	}
-	for _, pod := range pods.Items {
-		if pod.Status.Host != nodeID {
-			continue
+	for i := range nodes.Items {
+		var gracePeriod time.Duration
+		var lastReadyCondition api.NodeCondition
+		node := &nodes.Items[i]
+		readyCondition := s.getCondition(node, api.NodeReady)
+		if readyCondition == nil {
+			// If ready condition is nil, then kubelet (or nodecontroller) never posted node status.
+			// A fake ready condition is created, where LastProbeTime and LastTransitionTime is set
+			// to node.CreationTimestamp to avoid handle the corner case.
+			lastReadyCondition = api.NodeCondition{
+				Type:               api.NodeReady,
+				Status:             api.ConditionUnknown,
+				LastProbeTime:      node.CreationTimestamp,
+				LastTransitionTime: node.CreationTimestamp,
+			}
+			gracePeriod = nodeStartupGracePeriod
+		} else {
+			// If ready condition is not nil, make a copy of it, since we may modify it in place later.
+			lastReadyCondition = *readyCondition
+			gracePeriod = nodeMonitorGracePeriod
 		}
-		glog.V(2).Infof("Delete pod %v", pod.Name)
-		if err := s.kubeClient.Pods(pod.Namespace).Delete(pod.Name); err != nil {
-			glog.Errorf("Error deleting pod %v: %v", pod.Name, err)
+
+		// Check last time when NodeReady was updated.
+		if s.now().After(lastReadyCondition.LastProbeTime.Add(gracePeriod)) {
+			// NodeReady condition was last set longer ago than gracePeriod, so update it to Unknown
+			// (regardless of its current value) in the master, without contacting kubelet.
+			if readyCondition == nil {
+				glog.V(2).Infof("node %v is never updated by kubelet")
+				node.Status.Conditions = append(node.Status.Conditions, api.NodeCondition{
+					Type:               api.NodeReady,
+					Status:             api.ConditionUnknown,
+					Reason:             fmt.Sprintf("Kubelet never posted node status"),
+					LastProbeTime:      s.now(),
+					LastTransitionTime: s.now(),
+				})
+			} else {
+				// Note here the out-dated condition can be the one posted by nodecontroller
+				// itself before. We keep posting the status to keep LastProbeTime fresh.
+				glog.V(2).Infof("node %v hasn't been updated for a while, last ready condition is %+v", node.Name, readyCondition)
+				readyCondition.Status = api.ConditionUnknown
+				readyCondition.Reason = fmt.Sprintf("Kubelet stopped posting node status")
+				readyCondition.LastProbeTime = s.now()
+				if lastReadyCondition.Status != api.ConditionUnknown {
+					readyCondition.LastTransitionTime = s.now()
+				}
+			}
+			_, err = s.kubeClient.Nodes().Update(node)
+			if err != nil {
+				glog.Errorf("error updating node %s: %v", node.Name, err)
+			}
 		}
-	}
 
+		if readyCondition != nil {
+			// Check eviction timeout.
+			if lastReadyCondition.Status == api.ConditionNone &&
```

Comment:
Ok, the above comments are all related here; so let's discuss here.

I think the tradeoff here is refreshing LastProbeTime and a better logic.  I was leaning towards an up-to-date status from enduser's point; it is always good to see fresh status, especially when kubelet updates NodeReady=Full periodically.  Another reason is that, the status we are posting, the Unknown status, it is updated by nodecontroller, and its LastProbeTime and LastTransitionTime is updated at *now*, not the previous value.  IMO, we are hacking the status to fulfill the better logic.

The two are the primary reasons I'm refreshing LastProbeTime.  I did realize the flipping problem, but that's rare and is one of the reasons I'm introducing the grace period.

Combined the two, I think node controller still needs to set LastTransitionTime to now() at least, and LastProbeTime set to previous time.  The suggested logic can be used here.  How's that sound?